pkg/controller/promotions/promotions.go

```diff
@@ -709,3 +704,14 @@ func calculateRequeueInterval(p *kargoapi.Promotion) time.Duration {
 	}
 	return defaultRequeueInterval
 }
+
+// MarkStepAbortedIfRunning updates the newStatus to mark the current step as
+// aborted if the promotion is running and the current step is running.
+func MarkStepAbortedIfRunning(newStatus *kargoapi.PromotionStatus, promo *kargoapi.Promotion, finishedAt *metav1.Time) {
+	if newStatus.Phase == kargoapi.PromotionPhaseRunning &&
+		int64(len(newStatus.StepExecutionMetadata)) == promo.Status.CurrentStep+1 &&
+		promo.Status.StepExecutionMetadata[promo.Status.CurrentStep].Status == kargoapi.PromotionStepStatusRunning {
+		newStatus.StepExecutionMetadata[promo.Status.CurrentStep].Status = kargoapi.PromotionStepStatusAborted
+		newStatus.StepExecutionMetadata[promo.Status.CurrentStep].FinishedAt = finishedAt
+	}
+}
```

Comment:
If the intention is to make this reusable in other contexts, then this isn't the right place for it, as it will force awkward dependencies on this package. This should probably be moved to `pkg/api`.